### PR TITLE
feat(multiple packages): digital twins refresh

### PR DESCRIPTION
**Pull request by YoDaMa**
Once I get approval on the code I'll clean up the requirement docs.

### CODE REVIEW

File: digitaltwins/device/src/digital_twin_client.ts [hunk 1]
```diff
@@ -155,24 +139,31 @@ export class DigitalTwinClient {
...
+   * @description                   Adds multiple interfaceInstances to the Digital Twin client.  This will not cause
    *                                any network activity.  This is a synchronous method.
-   * @param newInterfaceInstance            The object for a particular interfaceInstance.
+   * @param newInterfaceInstances   A single object or multiple objects for a particular interfaceInstance.
+   */
+  addInterfaceInstances(...args: BaseInterface[]) {
+    for (var i = 0; i < args.length; i++) {
+      this._addInterfaceInstance(args[i]);
+    }
+  }
```
Comment (anthonyvercolano) [inline digitaltwins/device/src/digital_twin_client.ts]: lol

**Comment by anthonyvercolano** [comment]
>             break;

Maybe we should excise this now.

---
Refers to: digitaltwins/device/src/digital_twin_client.ts:201 in 747d6e9. [](commit_id = 747d6e97e40af171570f3137a43f1ed17ce150ab, deletion_comment = False)

**Comment by anthonyvercolano** [comment]
> };

This is gone sadly, but, it simply must go.

---
Refers to: digitaltwins/samples/device/javascript/sample_device.js:61 in 747d6e9. [](commit_id = 747d6e97e40af171570f3137a43f1ed17ce150ab, deletion_comment = False)

File: digitaltwins/samples/device/javascript/sample_device.js [hunk 2]
```diff
@@ -157,35 +64,44 @@ const environmentalSensor = new EnvironmentalSensor('environmentalSensor', prope
 const deviceInformation = new DeviceInformation('deviceInformation');
 const modelDefinition = new ModelDefinition('urn_azureiot_ModelDiscovery_ModelDefinition', null, modelDefinitionHandler);
 
-const deviceClient = DeviceClient.fromConnectionString(process.env.DEVICE_CONNECTION_STRING, Mqtt);
-
 const capabilityModel = 'urn:azureiot:samplemodel:1';
```
Comment (anthonyvercolano) [inline digitaltwins/samples/device/javascript/sample_device.js]: I had hoped that we would deal with this on this go round.  Really want to plumb the code for getting the DCM through on the connect.

File: digitaltwins/device/src/digital_twin_client.ts (hunk 1 above)
Comment (YoDaMa) [reply digitaltwins/device/src/digital_twin_client.ts]: I mean... i kinda like how it reads

---
In reply to: [401161653](https://github.com/Azure/azure-iot-sdk-node/pull/738#discussion_r401161653) [](ancestors = 401161653)

**Comment by YoDaMa** [comment]
>             break;

good idea. I'll add this to the PR.

---
In reply to: [606829100](https://github.com/Azure/azure-iot-sdk-node/pull/738#issuecomment-606829100) [](ancestors = 606829100)

---
Refers to: digitaltwins/device/src/digital_twin_client.ts:201 in 747d6e9. [](commit_id = 747d6e97e40af171570f3137a43f1ed17ce150ab, deletion_comment = False)

**Comment by YoDaMa** [comment]
> };

???

---
In reply to: [606850427](https://github.com/Azure/azure-iot-sdk-node/pull/738#issuecomment-606850427) [](ancestors = 606850427)

---
Refers to: digitaltwins/samples/device/javascript/sample_device.js:61 in 747d6e9. [](commit_id = 747d6e97e40af171570f3137a43f1ed17ce150ab, deletion_comment = False)

File: digitaltwins/samples/device/javascript/sample_device.js [hunk 3]
```diff
@@ -157,35 +64,44 @@ const environmentalSensor = new EnvironmentalSensor('environmentalSensor', prope
 const deviceInformation = new DeviceInformation('deviceInformation');
 const modelDefinition = new ModelDefinition('urn_azureiot_ModelDiscovery_ModelDefinition', null, modelDefinitionHandler);
```
Comment (anthonyvercolano) [inline digitaltwins/samples/device/javascript/sample_device.js]: excise

File: digitaltwins/samples/device/javascript/sample_device.js [hunk 4]
```diff
@@ -157,35 +64,44 @@ const environmentalSensor = new EnvironmentalSensor('environmentalSensor', prope
...
-  await digitalTwinClient.register();
+  // mqtt is implied in this static method
+  digitalTwinClient = DigitalTwinClient.fromConnectionString(process.env.DEVICE_CONNECTION_STRING);
+
+
+  // TBC: Do we create these inline
+  await digitalTwinClient.addInterfaceInstances(
+    environmentalSensor,
+    deviceInformation,
+    modelDefinition
```
Comment (anthonyvercolano) [inline digitaltwins/samples/device/javascript/sample_device.js]: excise.

File: digitaltwins/device/src/digital_twin_client.ts (hunk 1 above)
Comment (anthonyvercolano) [reply digitaltwins/device/src/digital_twin_client.ts]: Oh I agree, it was just perfect.

---
In reply to: [401188298](https://github.com/Azure/azure-iot-sdk-node/pull/738#discussion_r401188298) [](ancestors = 401188298,401161653)

File: digitaltwins/samples/device/javascript/sample_device.js [hunk 5]
```diff
@@ -157,35 +64,44 @@ const environmentalSensor = new EnvironmentalSensor('environmentalSensor', prope
...
 const capabilityModel = 'urn:azureiot:samplemodel:1';
 
 async function main() {
-  const digitalTwinClient = new DigitalTwinClient(capabilityModel, deviceClient);
-  digitalTwinClient.addInterfaceInstance(environmentalSensor);
-  digitalTwinClient.addInterfaceInstance(deviceInformation);
-  digitalTwinClient.addInterfaceInstance(modelDefinition);
-  await digitalTwinClient.register();
+  // mqtt is implied in this static method
+  digitalTwinClient = DigitalTwinClient.fromConnectionString(process.env.DEVICE_CONNECTION_STRING);
```
Comment (anthonyvercolano) [inline digitaltwins/samples/device/javascript/sample_device.js]: >fromConnectionString [](start = 40, length = 20)

Should the first parameter be the capability model?

File: digitaltwins/device/src/digital_twin_client.ts [hunk 6]
```diff
@@ -589,157 +586,31 @@ export class DigitalTwinClient {
...
-    });
+    let transport;
+    if (ws) {
+      transport = MqttWs;
+    } else {
+      transport = Mqtt;
+    }
+
+    const client = Client.fromConnectionString(connStr, transport);
+    client.setOptions({ deviceCapabilityModel: '&digital-twin-model-id=' + capabilityModel });
```
Comment (anthonyvercolano) [inline digitaltwins/device/src/digital_twin_client.ts:612]: >'&digital-twin-model-id=' [](start = 47, length = 25)

I probably would have put this down into mqtt-base.

File: digitaltwins/samples/device/javascript/sample_device.js (hunk 2 above)
Comment (YoDaMa) [reply digitaltwins/samples/device/javascript/sample_device.js:40]: this is now plumbed in as an option.

---
In reply to: [401188026](https://github.com/Azure/azure-iot-sdk-node/pull/738#discussion_r401188026) [](ancestors = 401188026)

File: digitaltwins/samples/device/javascript/sample_device.js (hunk 5 above)
Comment (YoDaMa) [reply digitaltwins/samples/device/javascript/sample_device.js]: yes.

---
In reply to: [401191616](https://github.com/Azure/azure-iot-sdk-node/pull/738#discussion_r401191616) [](ancestors = 401191616)

File: digitaltwins/device/src/digital_twin_client.ts (hunk 6 above)
Comment (YoDaMa) [reply digitaltwins/device/src/digital_twin_client.ts:530]: ok.

---
In reply to: [401979303](https://github.com/Azure/azure-iot-sdk-node/pull/738#discussion_r401979303) [](ancestors = 401979303)

File: device/transport/mqtt/src/mqtt.ts [hunk 7]
```diff
@@ -686,12 +690,13 @@ export class Mqtt extends EventEmitter implements DeviceTransport {
...
-    /*Tests_SRS_NODE_DEVICE_MQTT_41_002: [The MQTT constructor shall append the productInfo to the `username` property of the `config` object.]*/
+    /*Codes_SRS_NODE_DEVICE_MQTT_41_002: [The MQTT constructor shall append the productInfo to the `username` property of the `config` object.]*/
+    /*Codes_SRS_NODE_DEVICE_MQTT_41_XXX: [For a Plug and Play Device the device capability model should be included as &digital-twin-model-id=<DEVICE’s DCM ID> after the api-version]*/
     let baseConfig: MqttBaseTransportConfig = {
       uri: 'mqtts://' + (credentials.gatewayHostName || credentials.host),
       username: credentials.host + '/' + clientId +
-        '/' + endpoint.versionQueryString() +
-        '&DeviceClientType=' + encodeURIComponent(this._userAgentString),
+        '/' + endpoint.versionQueryString()  + this._dcm +
+        '&DeviceClientType=' + encodeURIComponent(this._userAgentString), // TODO: need to update useragentString
```
Comment (anthonyvercolano) [inline device/transport/mqtt/src/mqtt.ts:699]: >// TODO: need to update useragentString [](start = 74, length = 39)

Don't like TODO.  Please explain.

File: digitaltwins/samples/device/javascript/sample_device.js [hunk 8]
```diff
@@ -127,65 +33,47 @@ const commandHandler = (request, response) => {
...
-  const digitalTwinClient = new DigitalTwinClient(capabilityModel, deviceClient);
-  digitalTwinClient.addInterfaceInstance(environmentalSensor);
-  digitalTwinClient.addInterfaceInstance(deviceInformation);
-  digitalTwinClient.addInterfaceInstance(modelDefinition);
-  await digitalTwinClient.register();
+  // mqtt is implied in this static method
+  digitalTwinClient = DigitalTwinClient.fromConnectionString(capabilityModel, process.env.DEVICE_CONNECTION_STRING);
+
+
+  // TBC: Do we create these inline
```
Comment (anthonyvercolano) [inline digitaltwins/samples/device/javascript/sample_device.js:47]: >// TBC: Do we create these inline [](start = 2, length = 33)

Rather not.

File: digitaltwins/samples/device/javascript/sample_device.js [hunk 9]
```diff
@@ -127,65 +33,47 @@ const commandHandler = (request, response) => {
...
+  await digitalTwinClient.addInterfaceInstances(
```
Comment (anthonyvercolano) [inline digitaltwins/samples/device/javascript/sample_device.js:48]: >await [](start = 2, length = 5)

This is a synchronous function or should be.  Don't see the need for await.

**Comment by anthonyvercolano** [comment]
>   private readonly _sdkInformation: SdkInformation  = new SdkInformation('urn_azureiot_Client_SDKInformation', 'urn:azureiot:Client:SDKInformation:1');

We need to check on this.  Can't remember whether this is being excised or whether we control sending this with an option.  Need to talk with John or look at the spec.

---
Refers to: digitaltwins/device/src/digital_twin_client.ts:117 in 3b9d4b4. [](commit_id = 3b9d4b49fd4c08325c8500786c15b0b1669f4226, deletion_comment = False)

File: digitaltwins/device/src/digital_twin_client.ts [hunk 10]
```diff
@@ -156,23 +133,32 @@ export class DigitalTwinClient {
   //
   private _twin: Twin;
   constructor(capabilityModel: string, client: Client) {
```
Comment (anthonyvercolano) [inline digitaltwins/device/src/digital_twin_client.ts:135]: >capabilityModel: string, [](start = 14, length = 24)

Gone right?

File: digitaltwins/device/src/digital_twin_client.ts [hunk 11]
```diff
@@ -156,23 +133,32 @@ export class DigitalTwinClient {
...
-    /* Codes_SRS_NODE_DIGITAL_TWIN_DEVICE_06_001: [Will throw `ReferenceError` if `capabilityModel` argument is falsy.] */
+    /* Codes_SRS_NODE_DIGITAL_TWIN_DEVICE_06_001: [Will throw `ReferenceError` if the constructor `capabilityModel` argument is falsy.] */
     if (!capabilityModel) throw new ReferenceError('capabilityModel must not be falsy');
```
Comment (anthonyvercolano) [inline digitaltwins/device/src/digital_twin_client.ts:137]: Gone right?

**Comment by anthonyvercolano** [comment]
>       'iothub-command-statuscode': statusCode argument of the update method

Did we check to see if these property names have changed?

---
Refers to: digitaltwins/device/src/digital_twin_client.ts:357 in 3b9d4b4. [](commit_id = 3b9d4b49fd4c08325c8500786c15b0b1669f4226, deletion_comment = False)

File: digitaltwins/device/src/digital_twin_client.ts [hunk 12]
```diff
@@ -554,192 +455,76 @@ export class DigitalTwinClient {
...
 
     return callbackToPromise((_callback) => {
-      if (!this._interfaceInstances[interfaceInstanceName].registered) {
-        return _callback(new Error(interfaceInstanceName + ' is not registered'));
-      } else {
-        /* Codes_SRS_NODE_DIGITAL_TWIN_DEVICE_06_038: [** Properties may invoke the method `report` with a value to produce a patch to the reported properties. **] */
-        let interfaceInstancePart = interfaceInstancePrefix + interfaceInstanceName;
-        let propertyContent: any = {
-          value: propertyValue
+      let iName = interfaceInstance.interfaceInstanceName;
```
Comment (anthonyvercolano) [inline digitaltwins/device/src/digital_twin_client.ts:458]: >let iName = interfaceInstance.interfaceInstanceName; [](start = 6, length = 52)

Only use this variable once?  No need for it then.  Also the indentation is wrong after this.

**Comment by anthonyvercolano** [comment]
>    * @param interfaceInstanceName               Name of the instance for this interface.

Realize you didn't do this, could we fix it though.


---
Refers to: digitaltwins/device/src/digital_twin_client.ts:334 in 3b9d4b4. [](commit_id = 3b9d4b49fd4c08325c8500786c15b0b1669f4226, deletion_comment = False)

File: digitaltwins/device/test/_digital_twin_client_test.js [hunk 13]
```diff
@@ -140,165 +197,151 @@ describe('Digital Twin Client', function () {
...
-      contentType: 'application/json'
-      contentEncoding: 'utf-8'
-      **]
-    */
-    describe('sends a correctly formatted registration event', function () {
-      let dtClient;
-      let registrationDeviceClient;
-      let fakeInterfaceInstance;
+    /* Tests_SRS_NODE_DIGITAL_TWIN_DEVICE_41_XXX: [Can accept a variable number of interfaces to add to the Digital Twin Device Client] */
+    it.only('throws an Error if interface contains an unknown Digital Twin property type', function () {
```
Comment (anthonyvercolano) [inline digitaltwins/device/test/_digital_twin_client_test.js:205]: >it.only( [](start = 4, length = 8)

Not much is going to run in this suite.

**Comment by anthonyvercolano** [comment]
>     /* Tests_SRS_NODE_DIGITAL_TWIN_DEVICE_06_001: [Will throw `ReferenceError` if `capabilityModel` argument is falsy.] */

This is gone.

---
Refers to: digitaltwins/device/test/_digital_twin_client_test.js:25 in 3b9d4b4. [](commit_id = 3b9d4b49fd4c08325c8500786c15b0b1669f4226, deletion_comment = False)

File: digitaltwins/device/test/_digital_twin_client_test.js [hunk 14]
```diff
@@ -5,9 +5,11 @@
 const assert = require('chai').assert;
 const EventEmitter = require('events');
 const sinon = require('sinon');
+const sandbox = require('sinon').createSandbox();
```
Comment (anthonyvercolano) [inline digitaltwins/device/test/_digital_twin_client_test.js:8]: Why go for this?  Don't recall it anywhere else.

File: digitaltwins/device/test/_digital_twin_client_test.js [hunk 15]
```diff
@@ -41,7 +43,62 @@ describe('Digital Twin Client', function () {
...
+      });
+    });
+
+    /* Tests_SRS_NODE_DEVICE_CLIENT_41_XXX: [The fromConnectionString method shall return a new instance of the Client object, as by a call to new Client(new Transport(...)).] */
+    it('returns an instance of DigitalTwinClient', function () {
+      const dtClient = DigitalTwinClient.fromConnectionString(fakeCapabilityModel, fakeConnStr);
+      assert.instanceOf(dtClient, DigitalTwinClient);
+    });
+
+    /* Tests_SRS_NODE_DEVICE_CLIENT_41_XXX: [The fromConnectionString method shall use the interal MQTT transport by default] */
```
Comment (anthonyvercolano) [inline digitaltwins/device/test/_digital_twin_client_test.js:80]: >interal [](start = 91, length = 7)

spelling here and there.

consider using the cspell extension.

File: digitaltwins/device/test/_digital_twin_client_test.js [hunk 16]
```diff
@@ -140,165 +197,151 @@ describe('Digital Twin Client', function () {
...
-            assert.strictEqual(err, sendError);
+            assert.strictEqual(err, twinError);
             return done();
           });
       });
     });
 
+    /* Tests_SRS_NODE_DIGITAL_TWIN_DEVICE_41_XXX: [] */
+    it('will enable methods for all commands in all interfaceInstances', function (done) {
+      const registrationDeviceClient = {
```
Comment (anthonyvercolano) [inline digitaltwins/device/test/_digital_twin_client_test.js:314]: This is not about commands.

File: digitaltwins/device/test/_digital_twin_client_test.js [hunk 17]
```diff
@@ -140,165 +197,151 @@ describe('Digital Twin Client', function () {
...
+      };
+      class FakeInterface extends BaseInterface {
+        constructor(name, propertyCallback, commandCallback) {
+          super(name, 'urn:contoso:com:something:1', propertyCallback, commandCallback);
+          this.temp = new Command();
+          this.writableProperty = new Property(true);
+        }
+      };
+      const fakeInterfaceInstanceA = new FakeInterface('A', sinon.stub(), () => {});
+      const dtClient = new DigitalTwinClient('urn:abc:1', registrationDeviceClient);
```
Comment (anthonyvercolano) [inline digitaltwins/device/test/_digital_twin_client_test.js:333]: >ew DigitalTwinClient('urn:abc:1', registrationDeviceClient); [](start = 24, length = 60)

Need to update all these constructors.

File: device/transport/mqtt/src/mqtt.ts (hunk 7 above)
Comment (YoDaMa) [reply device/transport/mqtt/src/mqtt.ts:699]: removing

---
In reply to: [402700895](https://github.com/Azure/azure-iot-sdk-node/pull/738#discussion_r402700895) [](ancestors = 402700895)

File: digitaltwins/device/test/_digital_twin_client_test.js (hunk 14 above)
Comment (YoDaMa) [reply digitaltwins/device/test/_digital_twin_client_test.js:8]: I wanted to use sandbox to make creating and resetting sinon stubs easier. This way I don't worry about restoring sinon stubs after we are done with them, they are sandboxed for each test.

---
In reply to: [403089018](https://github.com/Azure/azure-iot-sdk-node/pull/738#discussion_r403089018) [](ancestors = 403089018)

File: digitaltwins/device/test/_digital_twin_client_test.js [hunk 18]
```diff
@@ -374,140 +301,72 @@ describe('Digital Twin Client', function () {
...
-    const almostToTimeout = 29000;
-    const pushOverTheTimeout = 2000;
-    describe('Will timeout in ' + timeoutForDisconnects + ' seconds if the initial registration message disconnects the connection', () => {
-      let dtClient;
-      let registrationDeviceClient;
-      let fakeInterfaceInstance;
-
+    /* Tests_SRS_NODE_DIGITAL_TWIN_DEVICE_41_XXX: [] */
+    it('will enable methods for all commands in all interfaceInstances', function (done) {
+      const registrationDeviceClient = {
```
Comment (anthonyvercolano) [inline digitaltwins/device/test/_digital_twin_client_test.js:306]: This is actually the first test that will run for the propertyupdates.  Need to change the comment will enable methods for all commands.

File: digitaltwins/samples/device/typescript/src/sample_device.ts [hunk 19]
```diff
@@ -204,26 +78,24 @@ const exitHandler = (request: CommandRequest, response: CommandResponse) => {
...
-    await deviceInformation.manufacturer.report('Contoso Device Corporation');
-    await deviceInformation.model.report('Contoso 4762B-turbo');
-    await deviceInformation.swVersion.report('3.1');
-    await deviceInformation.osName.report('ContosoOS');
-    await deviceInformation.processorArchitecture.report('4762');
-    await deviceInformation.processorManufacturer.report('Contoso Foundries');
-    await deviceInformation.totalStorage.report('64000');
-    await deviceInformation.totalMemory.report('640');
+    dtClient.report(environmentalSensor, {state: true});
+    dtClient.report(deviceInformation, {
```
Comment (anthonyvercolano) [inline digitaltwins/samples/device/typescript/src/sample_device.ts:89]: Shouldn't these use .then or await?

File: digitaltwins/device/test/_digital_twin_client_test.js [hunk 20]
```diff
@@ -20,28 +22,73 @@ const fakeDeviceClient = {
 
 describe('Digital Twin Client', function () {
   describe('#constructor', () => {
-    /* Tests_SRS_NODE_DIGITAL_TWIN_DEVICE_06_001: [Will throw `ReferenceError` if `capabilityModel` argument is falsy.] */
-    [undefined, null, ''].forEach(function (capabilityModel) {
-      it('throws a ReferenceError if \'capabilityModel\' is ' + capabilityModel + '\'', function () {
+    /* Tests_SRS_NODE_DIGITAL_TWIN_DEVICE_06_002: [Will throw `ReferenceError` if the constructor `client` argument is falsy.] */
+    [undefined, null, ''].forEach(function (client) {
+      it('throws a ReferenceError if \'client\' is ' + client + '\'', function () {
```
Comment (anthonyvercolano) [inline digitaltwins/device/test/_digital_twin_client_test.js:27]: > ' [](start = 50, length = 2)

Make this be \''

File: digitaltwins/device/test/_digital_twin_client_test.js [hunk 21]
```diff
@@ -20,28 +22,73 @@ const fakeDeviceClient = {
...
+    const fakeCapabilityModel = 'urn:fake:1';
+    const fakeConnStr = 'HostName=host;DeviceId=id;SharedAccessKey=key';
+
+    afterEach(() => {
+      sandbox.restore();
+    });
+
+    /* Tests_SRS_NODE_DIGITAL_TWIN_DEVICE_41_XXX: [Will throw `ReferenceError` if `capabilityModel` argument is falsy.] */
+    [undefined, null, ''].forEach(function (falsyCapabilityModel) {
+      it('throws a ReferenceError if \'capabilityModel\' is ' + falsyCapabilityModel + '\'', function () {
```
Comment (anthonyvercolano) [inline digitaltwins/device/test/_digital_twin_client_test.js]: > ' [](start = 59, length = 2)

Make this be \''  <- That is two ticks NOT double quote.

File: digitaltwins/device/test/_digital_twin_client_test.js [hunk 22]
```diff
@@ -20,28 +22,73 @@ const fakeDeviceClient = {
...
         });
       });
     });
 
-    /* Tests_SRS_NODE_DIGITAL_TWIN_DEVICE_06_002: [Will throw `ReferenceError` if the constructor `client` argument is falsy.] */
-    [undefined, null, ''].forEach(function (client) {
-      it('throws a ReferenceError if \'client\' is ' + client + '\'', function () {
+    /* Tests_SRS_NODE_DIGITAL_TWIN_DEVICE_41_XXX: [Will throw `ReferenceError` if the constructor `client` argument is falsy.] */
+    [undefined, null, ''].forEach(function (falsyConnStr) {
+      it('throws a ReferenceError if \'client\' is ' + falsyConnStr + '\'', function () {
```
Comment (anthonyvercolano) [inline digitaltwins/device/test/_digital_twin_client_test.js:46]: > ' [](start = 50, length = 2)

Again Make this be \''

File: digitaltwins/device/test/_digital_twin_client_test.js [hunk 23]
```diff
@@ -51,10 +98,10 @@ describe('Digital Twin Client', function () {
 
     /* Tests_SRS_NODE_DIGITAL_TWIN_DEVICE_06_003: [Will throw `ReferenceError` if the `newInterfaceInstance` argument is falsy.] */
     [undefined, null, ''].forEach(function (newInterfaceInstance) {
-      const dtClient = new DigitalTwinClient('urn:abc:1', fakeDeviceClient);
+      const dtClient = new DigitalTwinClient(fakeDeviceClient);
       it('throws a ReferenceError if \'newInterfaceInstance\' is ' + newInterfaceInstance + '\'', () => {
```
Comment (anthonyvercolano) [inline digitaltwins/device/test/_digital_twin_client_test.js:102]: > ' [](start = 64, length = 2)

Make this be \''

File: digitaltwins/device/test/_digital_twin_client_test.js [hunk 24]
```diff
@@ -69,9 +116,9 @@ describe('Digital Twin Client', function () {
       };
       const badFakeInterfaceInstance = new BadFakeInterface('badFakeInterfaceInstance');
       it('throws a ReferenceError if \'newInterfaceInstance\' \'interfaceId\` property is ' + interfaceId + '\'', () => {
```
Comment (anthonyvercolano) [inline digitaltwins/device/test/_digital_twin_client_test.js:118]: >' [](start = 90, length = 1)

Make this be \''

File: digitaltwins/device/test/_digital_twin_client_test.js [hunk 25]
```diff
@@ -80,9 +127,9 @@ describe('Digital Twin Client', function () {
     [undefined, null, ''].forEach(function (interfaceInstanceName) {
       const badFakeInterfaceInstance = new FakeInterface(interfaceInstanceName);
       it('throws a ReferenceError if \'newInterfaceInstance\' \'interfaceInstanceName\` property is ' + interfaceInstanceName + '\'', () => {
```
Comment (anthonyvercolano) [inline digitaltwins/device/test/_digital_twin_client_test.js:129]: >' [](start = 100, length = 1)

Make this be \''

File: digitaltwins/device/src/digital_twin_client.ts [hunk 26]
```diff
@@ -74,8 +62,8 @@ interface CommandInformation {
  *
  * An array of these items will be created for each interfaceInstance.
  *
- * At registration time, the registration code will sweep through all of the
- * registered interfaceInstances and within each interfaceInstance will utilize the data here to
+ * When enablePropertyUpdates is called, the code will sweep through all of the
+ * added interfaceInstances and within each interfaceInstance will utilize the data here to
  * process each write enabled property.
  *
  * This are two separate things that must be done post registration.
```
Comment (anthonyvercolano) [inline digitaltwins/device/src/digital_twin_client.ts:69]: >registration [](start = 55, length = 12)

enablement

**Comment by anthonyvercolano** [comment]
>             // We save this in a structure created for this interfaceInstance so that at registration time we can enable the IoT Hub device method handler.

enable?

---
Refers to: digitaltwins/device/src/digital_twin_client.ts:176 in 57a6618. [](commit_id = 57a6618d81b27fb6acd0b1332f68b805e2af1af0, deletion_comment = False)

**Pull request closed without merging**